### PR TITLE
ISSUE-2478: Transform like pattern to regex

### LINKED ISSUE #2478: The result of `select '%' like '\%'` is wrong
**Summary**

Like expression get wrong result.

In `Databend`:
```
mysql> select '%' like '\%';
+-----------------+
| ('%' LIKE '\%') |
+-----------------+
|               0 |
+-----------------+
1 row in set (0.01 sec)
Read 1 rows, 1 B in 0.003 sec., 302.44 rows/sec., 302.44 B/sec.

mysql> select '\%' like '\%';
+------------------+
| ('\%' LIKE '\%') |
+------------------+
|                1 |
+------------------+
1 row in set (0.01 sec)
Read 1 rows, 1 B in 0.004 sec., 281.96 rows/sec., 281.96 B/sec.
```

In `MySQL`:
```
mysql> select '%' like '\%';
+---------------+
| '%' like '\%' |
+---------------+
|             1 |
+---------------+
1 row in set (0.05 sec)

mysql> select '\%' like '\%';
+----------------+
| '\%' like '\%' |
+----------------+
|              0 |
+----------------+
1 row in set (0.04 sec)
```

In `Clickhouse`:
```
d95498e21498 :) select '%' like '\%';

SELECT '%' LIKE '\\%'

Query id: 982f747b-c52c-42bc-b7f2-eb62418f7dae

┌─like('%', '\\%')─┐
│                1 │
└──────────────────┘

1 rows in set. Elapsed: 0.083 sec. 

d95498e21498 :) select '\%' like '\%';

SELECT '\\%' LIKE '\\%'

Query id: 9725fb28-ec8b-4e53-a87f-96e810e195f7

┌─like('\\%', '\\%')─┐
│                  0 │
└────────────────────┘

1 rows in set. Elapsed: 0.084 sec.
```

**How to reproduce**

**Error message and/or stacktrace**

**Additional context**

### PR DESCRIPTION
**Pull request by zhyass**
I hereby agree to the terms of the CLA available at: https://databend.rs/policies/cla/

## Summary

1. Use backslash to escape the regex special character.
2. Process the backslash escape.

e.g. 
'Hello\\._World%\\%' tranform to '^Hello\\\\\\..World.*%$'.

## Changelog

- Bug Fix

## Related Issues

Fixes #2478 

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

**Comment by zhyass** [comment]
PTAL @jorgecarleitao

TODO: 
Unit Tests
Stateless Tests

File: common/datavalues/src/arrays/ops/like.rs [hunk 1]
```diff
@@ -112,6 +112,7 @@ fn is_like_pattern_escape(c: u8) -> bool {
     c == b'%' || c == b'_' || c == b'\\'
 }
 
+#[derive(Clone, Debug, PartialEq)]
```
Comment (jorgecarleitao) [inline common/datavalues/src/arrays/ops/like.rs:115]:
```suggestion
#[derive(Clone, Debug, PartialEq, Eq, Copy, Hash)]
```

since this is a simple enum.

**Review by jorgecarleitao — APPROVED** [review]
Looks great.

**Comment by jorgecarleitao** [comment]
oops, I did not intend to get this merged :/

File: common/datavalues/src/arrays/ops/like.rs (hunk 1 above)
Comment (zhyass) [reply common/datavalues/src/arrays/ops/like.rs:115]: > since this is a simple enum.

got it

**Pull request merged**